### PR TITLE
Fix race accessing plugin storage map

### PR DESCRIPTION
**Pull request by cpuguy83**
`plugins.GetAll()` was not locking the plugin map when checking if a
plugin exists, this can cause a race and potentially a panic if another
goroutine is trying to load a plugin into the map at the same time.

Also fixes a race during activation where a plugin inserts itself into
the plugin map but does not check if something else is already there.
This is already checked before trying to activate the plugin, however
the map lock is not held for this entire period, so other plugins may be
loaded during this time.
To fix, before inserting the plugin into the map, check if one with the
same name already exists and use that instead.

### CODE REVIEW

**Comment by cpuguy83** [comment]
ping @tiborvass @vieux 
We may want to take at least the map access issue as this could cause a panic.
I'm surprised one hasn't been reported already.

**Comment by tiborvass** [comment]
LGTM

File: pkg/plugins/plugins.go [hunk 1]
```diff
@@ -298,7 +302,10 @@ func GetAll(imp string) ([]*Plugin, error) {
 	chPl := make(chan *plLoad, len(pluginNames))
 	var wg sync.WaitGroup
 	for _, name := range pluginNames {
-		if pl, ok := storage.plugins[name]; ok {
+		strage.Lock()
```
Comment (AkihiroSuda) [inline pkg/plugins/plugins.go]: strage -> storage

**Comment by cpuguy83** [comment]
Sorry for the delay, this is now fixed.

**Review by AkihiroSuda — APPROVED** [review]
LGTM

**Comment by thaJeztah** [comment]
cherry picked into 1.13 through https://github.com/docker/docker/issues/29895

**Pull request merged**